sort/mergeSort.go

```diff
@@ -2,204 +2,157 @@ package sort
-		if done {
-			done = true
-			break
-		} else if curr != nil {
-			valueToAdd := curr.Copy()                  // Copy the curr value to a new memory address
-			chunk.Push((valueToAdd).(MergeSortSingle)) // Push the copied pointer to the slice
-		}
-	}
-	return chunk, done
+// GoExternalMergeSort inputs generic data records from a channel and a function to compare multiple data records
+// and outputs a channel of sorted data records. This function sorts externally via merge sort and therefore is
+// appropriate for very large amounts of data.
+func GoExternalMergeSort[E any](data <-chan E, recordsPerTmpFile int, less func(a, b E) bool) <-chan E {
+	out := make(chan E, 1000)
+	go mergeSort(data, less, out, recordsPerTmpFile)
+	return out
 }
 
-func mergeChunks(tmpFiles []string, outFilename string, filetype string, sortType string) {
-	fileReaders := make([]*fileio.EasyReader, len(tmpFiles))
-	var done bool
+func mergeSort[E any](data <-chan E, less func(a, b E) bool, out chan<- E, recordsPerTmpFile int) {
+	var err error
 
-	// This memory address map will store the memory address of each of the input variables
-	// and key them to the origin file reader. The memory addresses for each file is kept static
-	// throughout the heap loop below and therefore can be used as a stable key to determine
-	// the file of origin
-	memoryAddressMap := make(map[interface{}]*fileio.EasyReader)
+	// write records from data channel into multiple temporary files (chunks)
+	// that are each independently sorted using the input less function
+	var tmpFiles []*os.File
+	tmpFiles = chunkData(data, recordsPerTmpFile, less)
 
-	priorityQueue, curr := chooseDataType(filetype, sortType)
+	// a bit of hacking to make types match so that we can use a generic method
+	lessFunc = func(a, b any) bool {
+		return less(*a.(*E), *b.(*E))
+	}
 
-	for i := 0; i < len(tmpFiles); i++ {
-		fileReaders[i] = fileio.EasyOpen(tmpFiles[i])
-		done = curr.NextRealRecord(fileReaders[i])
-		if done {
-			continue
-		}
-		valueToAdd := curr.Copy()                   // Copy the curr value to a new memory address
-		writeVal := (valueToAdd).(MergeSortSingle)  // Assert the type of the empty interface valueToAdd
-		priorityQueue.Push(writeVal)                // Push the copied pointer to the heap
-		memoryAddressMap[writeVal] = fileReaders[i] // Key the memory address to the origin file reader
+	// This map will store the memory address of each index in a buffer slice (one per file)
+	// and key them to the decoder for the origin file. The memory addresses for the buffer
+	// is kept static throughout the heap loop below and therefore can be used as a stable key
+	// to determine the file of origin
+	memoryAddressMap := make(map[*E]*gob.Decoder)
+
+	curr := make([]*E, len(tmpFiles)) // this functions as a buffer for hb with static memory addresses
+	for i := range curr {
+		curr[i] = new(E) // allocate mem for curr pointers
+	}
+	hb := make(heapBuf, len(tmpFiles))
```

Comment:
I'm going to take notes on this because this type of make() works, but when I tried to make or use a variable with a generic type I got a "can't use generic type without instantiation" error

Reply:
I am still learning generics so I am not positive but I suspect the compiler makes exceptions for the `any` type. In most cases you need to instantiate when using a generic type but `any` is an alias for an empty interface `interface{}` so maybe that is why this is ok?